text/3493-precise-pre-release-cargo-update.md

```diff
@@ -0,0 +1,129 @@
+
+```
+[dependencies]
+example = "1.0.0"
+```
+
+It is possible to update to `1.2.0-pre.0` because `1.2.0` is semver compatible with `1.0.0`
+
+```
+> cargo update -p example --precise 1.2.0-pre.0
+    Updating crates.io index
+    Updating example v1.0.0 -> v1.2.0-pre.0
+```
+
+It is not possible to update to `2.0.0-pre.0` because `2.0.0` is not semver compatible with `1.0.0`
+
+```
+> cargo update -p example --precise 2.0.0-pre.0
+    Updating crates.io index
+error: failed to select a version for the requirement `example = "^1"`
+candidate versions found which didn't match: 2.0.0-pre.0
+location searched: crates.io index
+required by package `tmp-oyyzsf v0.1.0 (/home/ethan/.cache/cargo-temp/tmp-OYyZsF)`
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Consider this table where `a.b.c` is compatible with `x.y.z` and `x.y.z > a.b.c`
+
+| Cargo.toml spec | Cargo.lock version | Target version | Selected by cargo update  | Selected by cargo update --precise  |
+| --------------- | ------------------ | -------------- | ------------------------- | ----------------------------------- |
+| `a.b.c`         | `a.b.c`            | `x.y.z`        | ✅                        | ✅                                  |
+| `a.b.c`         | `a.b.c`            | `x.y.z-pre.0`   | ❌                        | ✅                                  |
+| `a.b.c`         | `x.y.z-pre.0`       | `x.y.z-pre.1`   | ❌                        | ✅                                  |
+| `a.b.c-pre.0`    | `a.b.c-pre.0`       | `a.b.c-pre.1`   | ✅¹                       | ✅                                  |
+| `a.b.c-pre.0`    | `a.b.c-pre.0`       | `x.y.z`        | ✅¹                       | ✅                                  |
+| `a.b.c`         | `a.b.c`            | `a.b.c-pre.0`   | ❌                        | ❌                                  |
+
+✅: Will upgrade
+
+❌: Will not upgrade
+
+¹For backwards compatibility with Cargo's current behaviour (see [RFC: Precise Pre-release Deps](https://github.com/rust-lang/rfcs/pull/3263))
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Pre-release versions are not easily auditable when they are only specified in the lock file.
+  A change that makes use of a pre-release version may not be noticed during code review as reviewers don't always check for changes in the lock file.
+- Library crates that require a pre-release version are not well supported since their lock files are ignored by their users (see [future-possibilities])
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The main alternative to this would be to accept that pre-release versions are not very usable and discourage their use.
+Cargo overrides can be used instead using `[patch]`.
+These provide a similar experience to pre-releases, however, they require that the library's code is somehow vendored outside of the registry, usually with git.
```

Comment:
There [has long been discussion](https://github.com/rust-lang/cargo/issues/9227) about allowing `[patch]` sections to point at crates published on a registry. Recently, the cargo team has started being interested in seeing progress in that direction. If this works out at the technical level and the cargo team except it this would automatically create an alternative for prerelease testing. Instead of publishing `foo@1.1.0-pre1` and asking people to manually update, the library could publish `foo-next@1.1.0` and ask people to try it using patch.

This suggests an extension to `[patch]` to allow "I want to patch foo@1.0.0 with this git dependency, *but ignore the version in git and treated as 1.1.0*". This (has long been requested)[https://github.com/rust-lang/cargo/issues/5640] for purposes of checking if a change is actually breaking. But could be used to solve prerelease testing by having the library publish `foo@1.1.0-pre1` and have the user patch it in as `foo@1.1.0`. We may even be want to consider magically doing this. "If you have a patch section pointed at a package, then cargo ignores the pre-release portion of that packages version".

Reply:
Thanks for pointing this out. I had not seen that discussion.

I've tried to represent it here https://github.com/rust-lang/rfcs/pull/3493/commits/c1fe92034c567f42e3e1293bb110ec618774a9d5 let me know if more needs to be added.